### PR TITLE
Add server-side lookup throttling

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

Sometimes, broker requires a way to throttle newly incoming traffic and requires to restrict number of incoming concurrent lookup requests. 

### Modifications

- Add broker side throttling to serve max allowed concurrent lookup request
- Update value of maxConcurrentLookupRequest dynamically at broker

### Result

Broker can have capability to throttle newly incoming traffic by restricting number of concurrent lookup request.

### CODE REVIEW

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -171,24 +182,33 @@ protected void handlePartitionMetadataRequest(CommandPartitionedTopicMetadata pa
...
-                                ex.getMessage(), requestId));
-                    } else {
-                        log.warn("Failed to get Partitioned Metadata [{}] {}: {}", remoteAddress, topic,
-                                ex.getMessage(), ex);
-                        ctx.writeAndFlush(Commands.newPartitionMetadataResponse(ServerError.ServiceNotReady,
-                                ex.getMessage(), requestId));
-                    }
-                    return null;
-                });
+        if (service.getLookupRequestSemaphore().tryAcquire()) {
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: To avoid indenting the whole section, what about doing: 

```java
if (!service.getLookupRequestSemaphore().tryAcquire()) {
    /// deal with error
    return;
}

// continue normal
```

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java [hunk 2]
```diff
@@ -171,24 +182,33 @@ protected void handlePartitionMetadataRequest(CommandPartitionedTopicMetadata pa
...
+                                        ex.getMessage(), ex);
+                                ctx.writeAndFlush(Commands.newPartitionMetadataResponse(ServerError.ServiceNotReady,
+                                        ex.getMessage(), requestId));
+                            }
+                        }
+                        service.getLookupRequestSemaphore().release();
+                        return null;
+                    });
+        } else {
+            log.warn("[{}] Failed Partition-Metadata lookup due to too many lookup-requets {}", remoteAddress, topic);
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: log at debug level

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 3]
```diff
@@ -820,6 +843,39 @@ public String generateUniqueProducerName() {
     public AuthenticationService getAuthenticationService() {
         return authenticationService;
     }
+    
+    private int getLookupRequestPermits() {
+        int pendingLookupRequest = pulsar.getConfiguration().getMaxConcurrentLookupRequest();
+        try {
+            Optional<Map<String, String>> data = dynamicConfigurationCache.get(LOOKUP_THROTTLING_PATH);
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Dynamic configuration of the parameter should go in separate PR

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java (hunk 1 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/ServerCnx.java]: Actually, here I have replaced ```thenAccept/exceptionally``` with ```handle(data, ex)``` so, don't have to keep duplicate ```service.getLookupRequestSemaphore().release()```. So, this change anyway changed entire section so, kept into ```if/else```.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java (hunk 3 above)
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: is that fine if we remove this configuration-update once other PR is created?

Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Either that, or just use the value from config file as it is (like any other config parameter) and then we can open a separate issue to discuss the broader dynamic config approach.

Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: yes, we can read from config-file but part of this PR also requires to test updating semaphore at runtime and if config value gets changed then we are also updating ```lookupRequestSemaphore``` so, even with Dynamic-config-change we might need this listener logic and we have to provide a way to inject listener on config value change in dynamic-config-PR.
So, should we keep this PR open until we come up with the approach as anyway it should be part of 1.17 release.

Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: Sure, that makes sense

**Comment by rdhabalia** [comment]
@merlimat rebased this one with the #186 changes.

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 4]
```diff
@@ -81,6 +81,7 @@ enum ServerError {
     TopicNotFound = 11; // Topic not found
     SubscriptionNotFound = 12; // Subscription not found
     ConsumerNotFound = 13; // Consumer not found
+    TooManyRequest = 14; // Error with too many simultaneously request
```
Comment (merlimat) [inline pulsar-common/src/main/proto/PulsarApi.proto]: Minor thing: this should be plural: `TooManyRequests`
